### PR TITLE
Port behaviours to filters

### PR DESCRIPTION
**Pull request by gunndabad**
Some time ago in a tech chat we spoke about moving cross-cutting things out of MediatR behaviors and into MVC filters. We now have a concrete reason to do so: we need the filters to be able to run on actions that don't use MediatR (e.g. actions that return only content).

This ports all except one of the current behaviors to filters. The final remaining one is a little more complicated and will be done separately.

### CODE REVIEW

File: src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs [hunk 1]
```diff
@@ -0,0 +1,20 @@
+﻿using Dfc.CourseDirectory.WebV2.Security;
+using Microsoft.AspNetCore.Mvc.Filters;
+using Microsoft.Extensions.DependencyInjection;
+
+namespace Dfc.CourseDirectory.WebV2.Filters
+{
+    public class AuthorizeAdminAttribute : ActionFilterAttribute
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs:7]: Why not derivative of `IAuthorizationFIlter`?

File: src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeCourseAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+            }
+
+            if (IsAuthorized())
+            {
+                var providerInfo = await providerInfoCache.GetProviderInfo(providerId.Value);
+                providerContextProvider.SetProviderContext(new ProviderContext(providerInfo));
+
+                await next();
+            }
+            else
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeCourseAttribute.cs:47]: Invert condition to remove unnecessary else

File: src/Dfc.CourseDirectory.WebV2/Filters/RestrictProviderTypesAttribute.cs [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+
+            // Previous filter has already checked this is set
+            var providerContext = providerContextProvider.GetProviderContext();
+
+            if ((providerContext.ProviderInfo.ProviderType & PermittedProviderTypes) == 0)
+            {
+                // The active provider doesn't have any of the permitted provider types
+                throw new NotAuthorizedException();
+            }
+            else
```
Comment (jameslawesynergy) [inline src/Dfc.CourseDirectory.WebV2/Filters/RestrictProviderTypesAttribute.cs:51]: Unnecessary else

**Review by jameslawesynergy — APPROVED** [review]

File: src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs (hunk 1 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs:7]: Relies on `NotAuthorizedException` being caught and handled by an exception filter. Exception filters don't catch errors from `IAuthorizationFilter`s, it seems.

File: src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeCourseAttribute.cs (hunk 2 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeCourseAttribute.cs:47]: Done

File: src/Dfc.CourseDirectory.WebV2/Filters/RestrictProviderTypesAttribute.cs (hunk 3 above)
Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Filters/RestrictProviderTypesAttribute.cs:51]: Done

File: src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs (hunk 1 above)
Comment (jameslawesynergy) [reply src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs:7]: Why does it need to throw a `NotAuthorizedException`? Can it not just return the response from the filter?

Comment (gunndabad) [reply src/Dfc.CourseDirectory.WebV2/Filters/AuthorizeAdminAttribute.cs:7]: It could but then we have the logic in two places instead of one.

**Pull request merged**